src/ray/raylet/node_manager.cc

```diff
@@ -541,6 +541,35 @@ ray::Status NodeManager::RegisterGcs() {
         RayConfig::instance().raylet_check_gc_period_milliseconds(),
         "NodeManager.CheckGC");
   }
+  // Raylet periodically check whether it's alive in GCS.
+  // For failure cases, GCS might think this raylet dead, but this
+  // failure still think it's alive.
+  periodical_runner_.RunFnPeriodically(
+      [this] {
+        // Flag to see whether a request is running.
+        static bool checking = false;
+        if (checking) {
+          return;
+        }
+        checking = true;
+        RAY_CHECK_OK(gcs_client_->Nodes().AsyncCheckSelfAlive([this](auto status,
+                                                                     auto alive) mutable {
+          if (status.ok()) {
+            if (!alive) {
+              // GCS think this raylet is dead. Set the raylet unhealthy.
+              RAY_LOG(WARNING)
+                  << "GCS consider this node to be dead. Set the status to be unhealthy";
+              node_manager_server_.GetServer().GetHealthCheckService()->SetServingStatus(
+                  self_node_id_.Hex(), false);
+              node_manager_server_.GetServer().GetHealthCheckService()->SetServingStatus(
+                  false);
+            }
+            checking = false;
+          }
```

Comment:
what if status is not okay? We just retry forever?

Reply:
Yes, if GCS is down for a long time, raylet will crash. 
Status not ok => GCS not reachable => Raylet crash eventually